### PR TITLE
Adding W&B (wandb) experiment tracking

### PR DESCRIPTION
**Pull request by kldarek**
# What does this PR do?
This PR adds a minimalistic W&B integration to track experiment configs, training and evaluation loss, perplexity, and system metrics like GPU/CPU utilization. 

## Feature/Issue validation/testing
Tested in single GPU and single node - multiple GPU setup: https://wandb.ai/darek-kleczek/llama-recipes-wandb

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/llama-recipes/blob/main/CONTRIBUTING.md#pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes?  
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Review by morganmcg1 — APPROVED** [review]
Few small comments, after that lgtm!

File: src/llama_recipes/finetuning.py [hunk 1]
```diff
@@ -45,12 +45,28 @@
...
+        raise ImportError(
+            "You are trying to use wandb which is not currently installed. "
+            "Please install it using pip install wandb"
+        )
+    from llama_recipes.configs import wandb_config as WANDB_CONFIG
+    wandb_config = WANDB_CONFIG()
+    wandb_entity = None if wandb_config.wandb_entity == 'none' else wandb_config.wandb_entity
+    update_config(wandb_config, **kwargs)
+    run = wandb.init(project=wandb_config.wandb_project, entity=wandb_entity)
+    run.config.update(train_config)
```
[the remark this answers is not in the document]
Comment (kldarek) [reply src/llama_recipes/finetuning.py:61]: The alternative is to be a bit more verbose - convert both configs into dicts, merge them and pass into config in `wandb.init`:
```
# Convert dataclass instances to dictionaries
dict_train_config = dataclasses.asdict(train_config)
dict_fsdp_config = dataclasses.asdict(fsdp_config)

# Merge the dictionaries
combined_config = {**dict_train_config, **dict_fsdp_config}
run = wandb.init(**init_dict, config=combined_config)
```
I kept the shorter code but happy to change if the above is clearer!

File: README.md [hunk 2]
```diff
@@ -177,6 +177,15 @@ sbatch multi_node.slurm
 ```
 You can read more about our fine-tuning strategies [here](./docs/LLM_finetuning.md).
 
+## Weights & Biases Experiment Tracking
+
+You can enable [W&B](https://wandb.ai/) experiment tracking by using `enable_wandb` flag as below. You can change the project name and entity in `wandb_config`.
```
[the remark this answers is not in the document]
Comment (kldarek) [reply README.md]: updated to `use_wandb`

File: README.md [hunk 3]
```diff
@@ -194,7 +203,7 @@ This folder contains a series of Llama2-powered apps:
 # Repository Organization
 This repository is organized in the following way:
 
-[configs](src/llama_recipes/configs/): Contains the configuration files for PEFT methods, FSDP, Datasets.
+[configs](src/llama_recipes/configs/): Contains the configuration files for PEFT methods, FSDP, Datasets, W&B experiment tracking.
```
[the remark this answers is not in the document]
Comment (kldarek) [reply README.md]: change to Weights & Biases

File: src/llama_recipes/configs/wandb.py [hunk 4]
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
[the remark this answers is not in the document]
Comment (kldarek) [reply src/llama_recipes/configs/wandb.py:1]: I've seen this pattern in some other open PR's, but we can move the config arguments into training config instead. @HamidShojanazeri let us know your preference here!

File: src/llama_recipes/finetuning.py [hunk 5]
```diff
@@ -45,12 +45,28 @@
...
+    except ImportError:
+        raise ImportError(
+            "You are trying to use wandb which is not currently installed. "
+            "Please install it using pip install wandb"
+        )
+    from llama_recipes.configs import wandb_config as WANDB_CONFIG
+    wandb_config = WANDB_CONFIG()
+    wandb_entity = None if wandb_config.wandb_entity == 'none' else wandb_config.wandb_entity
+    update_config(wandb_config, **kwargs)
+    run = wandb.init(project=wandb_config.wandb_project, entity=wandb_entity)
```
[the remark this answers is not in the document]
Comment (kldarek) [reply src/llama_recipes/finetuning.py]: updated the code to take in all keyword arguments from `wandb_config` and pass them into `wandb.init`

File: src/llama_recipes/configs/wandb.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (HamidShojanazeri) [reply src/llama_recipes/configs/wandb.py:1]: @kldarek agreed, train config is a better place for that.

[the remark this answers is not in the document]
Comment (HamidShojanazeri) [reply src/llama_recipes/configs/wandb.py:1]: thanks @kldarek , @morganmcg1 for the PR and sorry for late response needed some alignments on our integrations with other ecosystem projects. 

I think current place looks fine as it seems with a good number of configs to set here, so placing them in training config might be confusing.

File: src/llama_recipes/finetuning.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (HamidShojanazeri) [reply src/llama_recipes/finetuning.py:62]: how about moving the ` setup_wandb()` to config_utils.py? it may further help to separation logic of the code.

**Review by HamidShojanazeri — APPROVED** [review]
Thanks @kldarek @morganmcg1 for the PR and sorry for the long delay to merge.
